packages/react/src/hooks/useVisibilityLoader.js

```diff
@@ -86,7 +79,7 @@ const useVisibilityLoader = (ref, options) => {
         observerRef.current = null;
       }
     };
-  }, [isVisible, ref, intersectionObserverOptions, intersectionCallback, hasMoreToLoad]);
+  }, [isVisible, ref, intersectionObserverOptions, hasMoreToLoad, isLoading, onVisible]);
```

Comment:
sorry, removed my comment. Wanted to modify it but got interrupted by the baby. I think only the non primitives like intersectionObserverOptions are problematic...

Reply:
just skip the lodash merge for those

Reply:
not sure if the default props are recreated on render...

Reply:
I'm having a sense of Deja Vu :)   Recently there was a need in Graphite for a data grid that supports the potential for infinite rows and columns.  And by that, it should perform well with large datasets both vertically and horizontally.  The developer went away (never talked to me) and came back a couple of weeks later with the finished solution.  I loaded it up with 200 rows and 100 columns and it crawled to a halt.  We finally talked and he went away and created what was needed and it's blazingly fast for both infinite rows and infinite columns.  Part of the issue is that sometimes I don't think devs are fully aware of what they are solving (and I've been there many times myself).

So let's look at this... what are we solving here... because I see the code and I see it trying to solve 2 things and neither of which is going to help the current table perform better. 

When I worked with the Graphite dev on building the data-grid, I sat down and drew up this diagram, so that they could better visualize what they were doing (or trying to do).

![image](https://user-images.githubusercontent.com/1056553/145999660-9ec4e8cc-8011-4c98-8721-caafcad2232b.png)

For grid the biggest performance bottleneck is going to be rendering and it's typically rendering information that is not visible.  So in the case of the graphite data grid it was rendering columns and cells for information that was not visible.

The other thing is that loading sequential data and rendering sequential data shouldn't be enforced.  Today, the Table has a dropdown for "Pages".  Imagine if when the table first loaded, and you click the dropdown it showed 1 and 2, and then later when you click 2, and then after it loaded, you clicked the dropdown and it showed, 1, 2,  and 3, you click 3, it loads, and then it shows 1, 2, 3, and 4.  If it worked like that you'd think that was a silly implementation of data paging.

Loading data via some loadMore really just implies a sequential data model and that limits how you can load the data.  Also, tying it to a visibility trigger might not be the best approach.  Fundamentally you are building an incremental loading interface in such a way because your components are built a certain way.  The data model of the table and datalist is flawed and I think these need to be addressed before implementing a visibility loader.  And I think that if the visibility load is just loading more data and that data, both horizontally and vertically is being rendered into the DOM when its not visible, then I don't think it's solving any problems with the table.  This where a data contract is really needed that all component can adhere to.  In the simplest form... you need...

* totalCount (how many items does it know about)
* isItemLoaded(index) (check if an item is loaded)
* itemsAreLoading (flag indicating that we are busy loading items)
* loadItems(start, size) (async function that loads a range of items)

This is the bare minimum, but, using this information you can...
* totalCount can allow us to create scrollbars that reflect the total amount of items
* isItemLoaded(index) can be used to check if an item is loaded and show some sort of loading indicator if not...
* itemsAreLoading property can be used to signal that the "loader" is busy loading items
* loadItems(start, size) can be used to get a range of items that you need to render

Using a data contract you can get items without actually managing them internally, but, this still doesn't solve the lazy rendering of item data.  ie, that is a completely separate solution and has nothing to do with loading data, it just about how we can only render into the DOM the elements we need, no matter how much data we might have access to.

So, I think that the first step, in my mind is really fixing the data model for the Table so that it is really a data model, and NOT an hybrid data/ui model as it stands today.  Data really needs to be separated from the UI model.

Reply:
Thanks, @stuckless. This is great feedback and I completely agree. That's what I'm working toward in these two PRs. They aren't finished solutions, but are steps toward a solution by building these new hooks while maintaining existing functionality, creating some potential examples, and gathering your feedback. I would appreciate you taking a look at the [useDataLoader hook](https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3140/files#diff-20e8db082c0a53fafbec9a5df135593fff9f6f00ebff08c10c0ea200930d0253R9) and [story](https://deploy-preview-3140--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-list--with-infinite-scroll-data-loader) in [this PR](https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3140). It's built on this PR, so you'll see many of the same changes, but just takes them a bit farther. It's using the IntersectionObserver to load/unload data as it becomes visible on the screen, doesn't worry about managing data at all, and leaves that up to the consumer in the onLoadData callback. That callback is passed the `pageSize` and `start` parameters you mentioned and needs to return `data`, `loading` and `hasMoreData` flags, but I can easily re-work that to be based off the `totalCount` variable you mention above.

Reply:
@kevinsperrine I think we need to decouple the concepts of lazily loading data and lazily rendering data.  These are 2 different things and they don't depend on each other.  But more importantly, I feel like the data model of the table should be refactored because right now, we are building "interfaces" that sort of work with the current data model, but, these really aren;t going to benefit me, personally.  I will look at the use data hook, but I feel like existing data loading is missing the important concept that data doesn't need to be loaded sequentially... and that UI components should not be managing holding on to references of complete data sets.